### ISSUE TITLE
[Cypher] REMOVE inside MERGE ON MATCH throws InvalidSyntaxException

### ISSUE THREAD

**Issue by ikwattro** [neo4j/neo4j#6172]
Hi,

First of all Happy New Year !!!!

Use case : Remove a label to a node inside a MERGE `ON MATCH`:

Based on the default console.neo4j.org graph : 

```
MERGE (n:Crew {name:"Morpheus"}) 
ON MATCH REMOVE n:Crew


Error: Invalid input 'R': expected whitespace, comment or SET (line 1, column 44 (offset: 43))
"MERGE (n:Crew {name:"Morpheus"}) ON  MATCH REMOVE n:Crew"
```

**Comment by thobe** [comment]
You are right that it probably would be more friendly to let the parser accept more things in `ON MATCH` and `ON CREATE`, and then raise a semantic error instead.

The sample query is obviously not valid, even if we were to allow more types of mutations in `ON MATCH` and `ON CREATE`, since it tries to remove one of the parameters used to match the node, thus violating the invariant of `MERGE`.

**Comment by ikwattro** [comment]
@thobe Thanks for your reply.

To be honest, the `obvious` part was not really there in our team, this is the first time we hear that you can not remove parameters used to match the node. Simply because this query doesn't fail :

```
Query:
MERGE (n:Crew { name:"Morpheus" })
ON MATCH SET n.name = NULL


Query took 93 ms and returned no rows. 
Updated the graph - set 1 property 
```

**Comment by thobe** [comment]
Yes, that one should fail as well, but doesn't.

**Comment by ikwattro** [comment]
Thanks, good to know then.

So it turns out to be more a feature request at the end, my initial application use case is the following : 

```
MERGE (n:PullRequest {zid: {zid} })
ON MATCH REMOVE n:OpenedPullRequest, n:ClosedPullRequest
```

**Comment by thobe** [comment]
That looks a lot more sensible, and I can understand the desire to want to do that.

You could of course just `REMOVE` those labels in both the case of the node being matched and in the case of it being created, the effect would be the same: that the node _does not_ have those labels, regardless of whether it had them before or not.

```
MERGE (n:PullRequest {zid: {zid} })
REMOVE n:OpenedPullRequest, n:ClosedPullRequest
```

**Comment by ikwattro** [comment]
Sure yes.

**Comment by mvoila** [comment]
I fall in the same use-case.
Is it implemented now?

**Comment by hvub** [comment]
This can be done through a property as follows:

```
CYPHER 25
MERGE (n:Crew {name:"Morpheus"}) 
  ON MATCH SET n.matched = true
  ON CREATE SET n.matched = false
LET nFiltered = CASE WHEN n.matched THEN n ELSE null END
REMOVE nFiltered:Crew
RETURN n, labels(n), nFiltered, n.matched
```
